x-pack/plugin/security/src/internalClusterTest/java/org/elasticsearch/xpack/security/authc/ApiKeyIntegTests.java

```diff
@@ -1522,7 +1526,122 @@ public void testUpdateApiKey() throws ExecutionException, InterruptedException,
+        assertNotNull(responseWithNoExpectedUpdates);
+        assertThat(responseWithNoExpectedUpdates.getUpdated(), empty());
+        assertEquals(apiKeyIds.size(), responseWithNoExpectedUpdates.getNoops().size());
+        assertThat(responseWithNoExpectedUpdates.getNoops(), containsInAnyOrder(apiKeyIds.toArray()));
+        assertThat(responseWithNoExpectedUpdates.getErrorDetails().keySet(), containsInAnyOrder(notFoundIds.toArray()));
+        for (String apiKeyId : apiKeyIds) {
+            final Map<String, Object> doc = getApiKeyDocument(apiKeyId);
+            expectRoleDescriptorsForApiKey("role_descriptors", newRoleDescriptors, doc);
+            expectRoleDescriptorsForApiKey("limited_by_role_descriptors", expectedLimitedByRoleDescriptors, doc);
+            expectMetadataForApiKey(newMetadata, doc);
+        }
+
+        // Check that bulk update works when all updates result in errors
+        getSecurityClient().invalidateApiKeys(apiKeyIds.toArray(new String[0]));
+        final BulkUpdateApiKeyRequest requestWithErrors = new BulkUpdateApiKeyRequest(
+            shuffledList(apiKeyIds),
+            newRoleDescriptors,
+            newMetadata
+        );
+
+        final BulkUpdateApiKeyResponse responseWithErrors = executeBulkUpdateApiKey(TEST_USER_NAME, requestWithErrors);
+
+        assertThat(responseWithErrors.getUpdated(), empty());
+        assertThat(responseWithErrors.getNoops(), empty());
+        assertThat(responseWithErrors.getErrorDetails().keySet(), containsInAnyOrder(apiKeyIds.toArray()));
+    }
+
+    public void testBulkUpdateApiKeysWithDuplicates() throws ExecutionException, InterruptedException {
+        final Tuple<List<CreateApiKeyResponse>, List<Map<String, Object>>> apiKeys = createApiKeys(
+            TEST_USER_NAME,
+            randomIntBetween(3, 5),
+            null
+        );
+        final List<String> apiKeyIds = apiKeys.v1().stream().map(CreateApiKeyResponse::getId).toList();
+        final List<RoleDescriptor> newRoleDescriptors = randomValueOtherThan(null, this::randomRoleDescriptors);
+        final Map<String, Object> newMetadata = randomValueOtherThan(null, ApiKeyTests::randomMetadata);
+        final List<String> idsWithDuplicates = shuffledList(Stream.concat(apiKeyIds.stream(), apiKeyIds.stream()).toList());
+        assertEquals(idsWithDuplicates.size(), apiKeyIds.size() * 2);
+
+        BulkUpdateApiKeyResponse response = executeBulkUpdateApiKey(
+            TEST_USER_NAME,
+            new BulkUpdateApiKeyRequest(idsWithDuplicates, newRoleDescriptors, newMetadata)
+        );
+
+        assertNotNull(response);
+        assertThat(response.getErrorDetails(), anEmptyMap());
+        final List<String> allIds = Stream.concat(response.getUpdated().stream(), response.getNoops().stream()).toList();
+        assertThat(allIds, containsInAnyOrder(apiKeyIds.toArray()));
+
+        // Check not found IDs reported once for duplicates
```

Comment:
Yup you're right. The comment is misleading but the test is still useful since it validates we get an entry for each error. Will fix the comment to accurately reflect the point of the test section.